relayer/chains/wasm/provider.go

```diff
@@ -569,42 +572,111 @@ func (p *Provider) fetchBlockMessages(ctx context.Context, heightInfo *types.Hei
 	var (
 		wg           sync.WaitGroup
 		messages     coreTypes.ResultTxSearch
-		messagesChan = make(chan *coreTypes.ResultTxSearch)
-		errorChan    = make(chan error)
+		messagesChan = make(chan *coreTypes.ResultTxSearch, len(p.eventList))
+		errorChan    = make(chan error, len(p.eventList))
 	)
 
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
 	for _, event := range p.eventList {
 		wg.Add(1)
-		go func(wg *sync.WaitGroup, searchParam types.TxSearchParam, messagesChan chan *coreTypes.ResultTxSearch, errorChan chan error) {
+		go func(wg *sync.WaitGroup, event sdkTypes.Event, searchParam types.TxSearchParam, messagesChan chan *coreTypes.ResultTxSearch, errorChan chan error) {
 			defer wg.Done()
-			searchParam.Events = append(searchParam.Events, event)
-			res, err := p.client.TxSearch(ctx, searchParam)
+
+			var (
+				localSearchParam = searchParam
+				localMessages    = new(coreTypes.ResultTxSearch)
+				zapFields        = []zap.Field{
+					zap.Uint64("start_height", localSearchParam.StartHeight),
+					zap.Uint64("end_height", localSearchParam.EndHeight),
+					zap.String("event", event.Type),
+				}
+			)
+			localSearchParam.Events = append(localSearchParam.Events, event)
+
+			err := retry.Retry(ctx, p.logger, func() error {
+				p.logger.Info("fetching block messages", zapFields...)
+				msgs, err := p.client.TxSearch(ctx, localSearchParam)
+				if err == nil {
+					p.logger.Info("fetched block messages", zapFields...)
+					localMessages = msgs
+				}
+				return err
+			}, zapFields)
 			if err != nil {
-				errorChan <- err
+				select {
+				case errorChan <- err:
+				default:
```

Comment:
any particular reason for using this non blocking send ?

Reply:
just to make things flexible.

Reply:
just to make things flexible.